File: ruby/tabulation/canConstruct.rb
def canConstruct(target, wordBank)
    table = Array.new(target.length+1, false)
    table[0] = true

    (target.length).times do |i|
        if (table[i] === true)
            wordBank.each do |word|
                if (target.slice(i, word.length) === word)
                    table[i+word.length] = true
                end
            end
        end
    end
    p table
    table[target.length]
end

p canConstruct('abcdefg', ['ab', 'bc', 'd', 'abcd', 'ef', 'lg'])

File: javascript/tabulation/canConstruct.js
const canConstruct = (targetString, wordBank) => {
    const table = Array(targetString + 1).fill(false);
    table[0] = true;

    for (let i=0; i <= targetString.length; i++) {
        if (table[i] === true) {
            for (let word of wordBank) {
                if (targetString.slice(i, i+word.length) === word) {
                    table[i+word.length] = true;
                }
            }
        }
    }
   return table[targetString.length];
};

console.log(canConstruct('abcdefg', ['ab', 'bc', 'd', 'abcd', 'ef', 'g']))

